app/routes/erddap/buoy.js

```diff
@@ -63,14 +63,30 @@ router.get(
   "/:source/query",
   ash(async (req, res) => {
     const ids = req.query.ids.split(",");
+    const withUnits = req.query.units === "true" ?? false;
 
     // check variables queried exist for this dataset, short circuit if they don't
     const queryVariables = req.query.variables.split(",");
     const datasetVariables =
-      mcache.get(`__express__/erddap/${req.params.source}/variables`) ?? [];
-    const variables = queryVariables
-      .filter((v) => datasetVariables.includes(v))
-      .filter((v) => !v.includes("Qualifiers"));
+      mcache.get(
+        `__express__/erddap/${req.params.source}/variables${
+          withUnits && "?units=true"
+        }`
+      ) ?? (await common.getVariables(req.datasetId, withUnits));
+
+    let variables;
+    if (withUnits) {
+      variables = queryVariables
+        .filter((v) =>
+          datasetVariables.map((variable) => variable.name).includes(v)
+        )
+        .filter((v) => !v.includes("Qualifiers"));
```

Comment:
there are about 20 or so variables in the biggest dataset, so it's not the most efficient, but also not that big.  I don't think it would be that much less readable.  I'll try it and stare at it and see how it looks!